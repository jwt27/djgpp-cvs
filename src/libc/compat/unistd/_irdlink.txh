@node __internal_readlink, io
@findex __internal_readlink
@subheading Syntax

@example
#include <libc/symlink.h>

int __internal_readlink(const char * path, int fhandle,
                        char * buf, size_t max)
@end example

@subheading Description
In general applications shouldn't call this function; use @code{readlink} 
instead (@pxref{readlink}).  However, there is one exception: if you have
a @code{FSEXT} @code{fstat} file handler, and do not want do anything special
about symlinks.  In this case you should call this function from your handler
to set properly @code{S_IFLNK} bit in @code{st_mode}.  This function operates
on either @var{path} @strong{or} @var{fhandle}. In any case, the other arg
should be set to @code{NULL} or 0.

@subheading Return Value

Number of copied characters; value -1 is returned in case of
error and @code{errno} is set.  When value returned is equal to 
@var{size}, you cannot determine if there was enough room to 
copy whole name.  So increase @var{size} and try again.

@subheading Portability

@portability !ansi, !posix

@subheading Example

@example
char buf[FILENAME_MAX + 1];
if (__internal_readlink(0, "/dev/env/DJDIR/bin/sh.exe",
                        buf, FILENAME_MAX) == -1)
   if (errno == EINVAL)
      puts("/dev/env/DJDIR/bin/sh.exe is not a symbolic link.");
@end example

