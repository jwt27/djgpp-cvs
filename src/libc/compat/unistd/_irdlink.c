/* Copyright (C) 2004 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2003 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2002 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2001 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2000 DJ Delorie, see COPYING.DJ for details */

#include <libc/stubs.h>
#include <libc/symlink.h>
#include <sys/fsext.h>
#include <libc/fsexthlp.h>
#include <dir.h>
#include <errno.h>
#include <fcntl.h>
#include <io.h>
#include <stdarg.h>
#include <stddef.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include "xsymlink.h"

int __internal_readlink(const char * __path, int __fhandle, char * __buf, 
                        size_t __max)
{
   ssize_t       bytes_read         = 0;
   char          buf[_SYMLINK_FILE_LEN];
   char        * data_buf;
   int           fd;
   int           ret;
   off_t         old_pos = 0;
   long          file_size = 0;

   /* Reject NULL and impossible arg combinations */
   if (!__buf || (__path && __fhandle) || !(__path || __fhandle))
   {
      errno = EINVAL;
      return -1;
   }

   /* Provide ability to hook symlink support */
   if (__path)
   {
      struct ffblk  file_info;
      if (__FSEXT_call_open_handlers_wrapper(__FSEXT_readlink, &ret,
					     __path, __buf, __max))
         return ret;
      /* We will check if file exists by the way */
      if (findfirst(__path, &file_info, FA_RDONLY|FA_ARCH))
      {
         errno = ENOENT;
         return -1;
      }
      file_size = file_info.ff_fsize;
   }
   else if (__fhandle)
   {
      __FSEXT_Function *func = __FSEXT_get_function(__fhandle);
      if (func)
      {
         int rv;
         if (__FSEXT_func_wrapper(func, __FSEXT_readlink, &rv, __path))
            return rv;
      }
      file_size = filelength(__fhandle);
   }

   /* Is symlink file size a fixed magic value? */
   if (file_size != _SYMLINK_FILE_LEN)
   {
      errno = EINVAL; /* Sad but true */
      return -1;
   }

   /* Now we check for special DJGPP symlink format */

   if (__fhandle)
   {
      /* Remember old file pos */
      old_pos = tell(__fhandle);
      if (old_pos == -1)
         return -1;
      lseek(__fhandle, 0, SEEK_SET);
      fd = __fhandle;
   }
   else
   {
      fd = _open(__path, O_RDONLY);
      if (fd < 0)
      {
         /* Retry with DENY-NONE share bit set. It might help in some cases
          * when symlink file is opened by another program. We don't try with
          * DENY-NONE set in the first _open() call, because it might fail under
          * some circumstances. For details, see Ralf Brown's Interrupt List,
          * description of INT 0x21, function 0x3D.
          */
         fd = _open(__path, O_RDONLY | SH_DENYNO);
         if (fd < 0)
            return -1; /* errno from _open() call */
      } 
   }

   bytes_read = _read(fd, &buf, _SYMLINK_FILE_LEN);

   if (__fhandle)
      lseek(__fhandle, old_pos, SEEK_SET);
   else
      _close(fd);

   if (bytes_read == -1)
      return -1; /* Return errno set by _read() (_close() in worse case) */
      
   /* Check for symlink file header */
   if (strncmp(buf, _SYMLINK_PREFIX, _SYMLINK_PREFIX_LEN))
   {
      if (!__fhandle)
         close(fd);
      errno = EINVAL;
      return -1;
   }
   
   data_buf = buf + _SYMLINK_PREFIX_LEN;
   bytes_read = strpbrk(data_buf, "\r\n") - data_buf;
   if( (unsigned)bytes_read > __max ) 
     bytes_read = __max;
   memcpy(__buf, data_buf, bytes_read);
   return bytes_read;
}

