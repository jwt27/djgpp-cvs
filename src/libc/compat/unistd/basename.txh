
@c ----------------------------------------------------------------------

@node basename, file system
@findex basename
@subheading Syntax

@example
#include <unistd.h>

char * basename (const char *fname);
@end example

@subheading Description

This function returns the @dfn{basename} of the file, which is the last
part of its full name given by @var{fname}, with the drive letter and
leading directories stripped off.  For example, the basename of
@code{c:/foo/bar/file.ext} is @code{file.ext}, and the basename of
@code{a:foo} is @code{foo}.  Trailing slashes and backslashes are
significant: the basename of @code{c:/foo/bar/} is an empty string after
the rightmost slash.

This function treats both forward- and backslashes like directory
separators, so it can handle file names with mixed styles of slashes.

@subheading Return Value

A pointer into the original file name where the basename starts.  Note
that this is @strong{not} a new buffer allocated with @code{malloc}.  If
@var{fname} is a NULL pointer, the function will return a NULL pointer.

@subheading Portability

@portability !ansi, !posix

@subheading Example

@example
  if (strcmp (basename (file_name), "gcc.exe") == 0)
    printf ("The file %s is the GNU C/C++ compiler\n", file_name);
@end example

