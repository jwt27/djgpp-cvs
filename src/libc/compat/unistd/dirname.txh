
@c ----------------------------------------------------------------------

@node dirname, file system
@findex dirname
@subheading Syntax

@example
#include <unistd.h>

char * dirname (const char *fname);
@end example

@subheading Description

This function returns the directory part of the argument @var{fname}
copied to a buffer allocated by calling @code{malloc}.  The directory
part is everything up to but not including the rightmost slash (either
forward- or backslash) in @var{fname}.  If @var{fname} includes a drive
letter but no slashes, the function will return @code{@var{x}:.} where
@var{x} is the drive letter.  If @var{fname} includes neither the drive
letter nor any slashes, @code{"."} will be returned.  Trailing slashes
are removed from the result, unless it is a root directory, with or
without a drive letter.

@subheading Return value

The directory part in malloc'ed storage, or a NULL pointer of either
there's not enough free memory, or @var{fname} is a NULL pointer.

@subheading Portability

@portability !ansi, !posix

@subheading Example

@example
 printf ("The parent of current directory is %s\n", 
         dirname (getcwd (0, PATH_MAX)));
@end example
