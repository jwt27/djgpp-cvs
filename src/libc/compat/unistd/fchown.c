/* Copyright (C) 2003 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2002 DJ Delorie, see COPYING.DJ for details */
#include <libc/stubs.h>
#include <stdarg.h>
#include <sys/fsext.h>
#include <libc/fsexthlp.h>
#include <io.h>
#include <unistd.h>

/* MS-DOS couldn't care less about file ownerships, so we 
   at least check if given handle is valid. */
 
int fchown(int fd, uid_t owner, gid_t group)
{
  __FSEXT_Function * func = __FSEXT_get_function(fd);
  if (func)
  {
    int rv;
    if (__FSEXT_func_wrapper(func, __FSEXT_fchown, &rv, fd, owner, group))
      return rv;
  }
  return (_get_dev_info(fd) == -1) ? -1 : 0;
}
