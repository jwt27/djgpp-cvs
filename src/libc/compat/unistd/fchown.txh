@node fchown, unix
@findex fchown
@subheading Syntax

@example
#include <unistd.h>

int fchown(int fd, uid_t owner, gid_t group);
@end example

@subheading Description

This function changes the ownership of the open file specified by @var{fd}
to the user ID @var{owner} and group ID @var{group}.

This function does almost nothing under MS-DOS: it just
checks if the handle @code{fd} is valid.  This function can 
be hooked by File System Extensions (@pxref{File System Extensions}).

@subheading Return Value

This function returns zero if the handle is valid, non-zero
otherwise.

@subheading Portability

@portability !ansi, !posix
