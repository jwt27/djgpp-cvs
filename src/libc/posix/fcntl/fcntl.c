/* Copyright (C) 2015 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2010 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2003 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2002 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2001 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1999 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1996 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1995 DJ Delorie, see COPYING.DJ for details */
#include <libc/stubs.h>
#include <dpmi.h>
#include <errno.h>
#include <stdarg.h>
#include <unistd.h>
#include <fcntl.h>
#include <io.h>
#include <sys/fsext.h>
#include <libc/fsexthlp.h>
#include <sys/movedata.h>
#include <libc/farptrgs.h>
#include <libc/dosio.h>
#include <libc/getdinfo.h>
#include <libc/fd_props.h>
#include <go32.h>


static unsigned long _get_sft_entry_ptr(int fd)
{
  __dpmi_regs regs;
  unsigned char ind;
  unsigned long es, di;


  /* Get the JFT entry address for this handle.  */
  regs.x.ax = 0x1220;
  regs.x.bx = fd;
  __dpmi_int(0x2f, &regs);


  if (regs.x.flags & 1) /* int2F/1220 returns CF set if bad input */
  {
    errno = __doserr_to_errno(regs.h.al);  /* And AL has the error number */
    return 0;
  }


  /* Get the SFT entry number for this handle.  */
  es = regs.x.es;
  di = regs.x.di;
  ind = _farpeekb(_dos_ds, es * 16 + di);


  if (ind < 0xff)
  {
    /* Now get the address of the entry.  */
    regs.x.ax = 0x1216;
    regs.x.bx = ind;
    __dpmi_int (0x2f, &regs);
  }
  else
    regs.x.flags |= 1;  /* Set carry flag to simulate error       */

  if (regs.x.flags & 1) /* int2F/1216 returns CF set if bad input */
  {
    errno = EBADF;      /* But no other information is returned   */
    return 0;
  }


  es = regs.x.es;
  di = regs.x.di;


  return es * 16 + di;
}


static int
inherit_bit_test (int fd, short dev_info)
{
  __dpmi_regs regs;
  short new_dev_info;


  dev_info |= _DEV_NO_INHERIT;
  regs.x.ax = 0x4401;
  regs.x.bx = fd;
  regs.x.dx = dev_info;
  __dpmi_int(0x21, &regs);

  new_dev_info = _get_dev_info(fd);

  /* If the dev info words are equal, then the documented
     interface can be used to set the inheritance bit.  */
  return (new_dev_info == dev_info) ? 1 : 0;
}

static int
_get_SHARE_status (void)
{
  __dpmi_regs regs;

  regs.x.ax = 0x1000;
  __dpmi_int(0x2f, &regs);

  /* If al == 0xff then SHARE is installed, otherwise not */
  return (regs.h.al == 0xff) ? 1 : 0;
}

static int
_fcntl_lk64(int fd, int cmd, struct flock64 *lock_r64)
{
  int ret = -1;
  long long int ret64 = -1L;
  offset_t pos, cur_pos, lock_pos, len;
  long long int flen;

  /* Is this a directory? If so, fail. */
  if (__get_fd_flags(fd) & FILE_DESC_DIRECTORY)
  {
    errno = EINVAL;
    return -1;
  }

  /* Check if SHARE is loaded */
  ret = _get_SHARE_status();

  if (!ret) /* Then SHARE is NOT loaded, just return success */
  {
    if (cmd == F_GETLK64) /* Then make believe the lock is available */
      lock_r64->l_type = F_UNLCK;

    return ret;
  }
  else ret = -1; /* Restore default value */

  cur_pos = llseek(fd, 0L, SEEK_CUR);
  if (cur_pos < 0)
    return -1;      /* Assumes llseek has set errno */

  lock_pos = llseek (fd, lock_r64->l_start, lock_r64->l_whence);
  if (lock_pos < 0)
    return -1;      /* Assumes llseek has set errno */

  len = lock_r64->l_len;


  /* If l_len is zero, then the lock is to be set from the position
     represented by l_start/l_whence to the current end-of-file.
  */
  if (len == 0L)
  {
    flen = lfilelength(fd); /* Hold lfilelength in case len < 0 */
    len = flen - lock_pos;  /* len > 0      lock_pos before EOF
                               len == 0     lock_pos at     EOF
                               len < 0      lock_pos after  EOF
                            */

    /* If now len < 0, then lock_pos is beyond EOF, and
       the code below will calculate the correct region.
    */
    if (len < 0L)
    {
      pos = flen;   /* Start at EOF == lfilelength(fd) */
      lock_pos = flen; /* Also set lock_pos for later  */
      len = -len;   /* Make the length positive        */
    }
  }


  /* Return to saved position */
  ret64 = llseek (fd, cur_pos, SEEK_SET);
  if (ret64 < 0L)
    return -1;      /* Assumes llseek has set errno */
  else
    ret64 = -1;     /* Restore default value */


  /* If l_len is positive, the area to lock is from l_start
     to l_start + l_len - 1. If l_len is negative, the area to lock is
     from l_start + len to l_start - 1.

     Note that if l_len == 0, then pos and len may have already
     been set above, and lock_pos may have been reset to flen.
  */
  if (len > 0L)
  {
    pos = lock_pos;
  }
  else
  {
    pos = lock_pos + len;
    len = -len;
  }


  /* DOS/Win9x only support write locks via int 21/5C, so
     all read lock requests are treated like write locks
  */

  if (lock_r64->l_type == F_UNLCK)
  {
    ret = _dos_unlk64(fd, pos, len);
    if (ret != 0)
    {
      _doserrno = ret;
      errno = __doserr_to_errno(ret);
      return -1;
    }
  }
  else if ((lock_r64->l_type == F_WRLCK) ||
           (lock_r64->l_type == F_RDLCK))
  {
    ret = _dos_lk64(fd, pos, len);
    if (cmd == F_GETLK64)
    {
      if (!ret)
      {
        _dos_unlk64(fd, pos, len);
        /* If no lock is found that would prevent a lock from
           being created, the lock type is set to F_UNLCK.
        */
        lock_r64->l_type = F_UNLCK;
      }
      else
      {
        /* If a lock is found then l_whence, l_start, and l_len
           should point to the area covered by the lock. But the
           file locking interface doesn't report where the
           existing lock is, so nothing more can be done.
        */
        _doserrno = ret;
        errno = __doserr_to_errno(ret);
        return -1;
      }
    }
    else
    {
      /* If F_SETLKW64 is set, wait for the lock to be released.  */
      if (cmd == F_SETLKW64 && ret)
      {
        while ((ret = _dos_lk64(fd, pos, len)))
          __dpmi_yield();
      }
      else if (ret)
      {
        _doserrno = ret;
        errno = __doserr_to_errno(ret);
        return -1;
      }
    }
  }
  if (ret)
    errno = ENOSYS;

  return ret;
}


int
fcntl(int fd, int cmd, ...)
{
  int tofd, open_max;
  va_list ap;
  __FSEXT_Function *func;
  short dev_info = _get_dev_info(fd);
  static int inherit_bit_visible = -1;
  int errno_save;

  /* Verify the descriptor is valid by retrieving
     the handle's device info word.  */
  if (dev_info == -1)
    return dev_info;


  /* Allow a fd to override with a FSEXT.  */
  func = __FSEXT_get_function(fd);
  if (func)
  {
    int rv;
    if (__FSEXT_func_wrapper(func, __FSEXT_fcntl, &rv, fd))
      return rv;
  }


  errno_save = errno;

  switch (cmd)
  {
    case F_DUPFD:
    {
      va_start(ap, cmd);
      tofd = va_arg(ap, int);
      va_end(ap);


      open_max = getdtablesize();
      while (tofd < open_max)
      {
        /* If unable to get the device info for the handle,
           then the handle is not active and it can be used.  */
        if (_get_dev_info(tofd) == -1)
          break;
        tofd++;
      }


      if (tofd >= open_max)
      {
        errno = EMFILE;
        return -1;
      }


      errno = errno_save;
      return dup2(fd, tofd);
    }


    case F_GETFD:
    {
       unsigned long entry_ptr;


      /* DOS only passes the first 20 handles to child programs.  In
         addition, handles 19 and 18 will be closed by the stub of the
         child program (if it is a DJGPP program).  */


      if (fd >= 18)
        return FD_CLOEXEC;


      /* Determine if the documented interface will allow twiddling with
         the inherit bit. If not, fallback to the undocumented one.  */
      if (inherit_bit_visible == -1)
        inherit_bit_visible = inherit_bit_test (fd, dev_info);

      if (!inherit_bit_visible)
      {
        entry_ptr = _get_sft_entry_ptr(fd);
        if (entry_ptr == 0)
        {
          /* The fd has already been validated, so reaching here means
             something is wrong with _get_sft_entry_ptr. */
          return -1;
        }
        /* Offset 5 in the SFT contains the device info word.  */
        dev_info = _farpeekw(_dos_ds, entry_ptr + 5);
      }
      return (dev_info & _DEV_NO_INHERIT) ? FD_CLOEXEC : 0;
    }


    case F_SETFD:
    {
      unsigned int flag;
      unsigned long entry_ptr = 0; /* shut up -Wall */
      __dpmi_regs regs;


      va_start (ap, cmd);
      flag = va_arg(ap, int);
      va_end(ap);


      /* DOS only passes the first 20 handles to child programs.  In
         addition, handles 19 and 18 will be closed by the stub of the
         child program (if it is a DJGPP program).  */
      if (fd >= 18)
      {
        if (flag & FD_CLOEXEC)
          return 0;
        else
        {
          errno = ENOSYS;
          return -1;
        }
      }


      /* Determine if the documented interface will allow twiddling with
         the inherit bit. If not, fallback to the undocumented one.  */
      if (inherit_bit_visible == -1)
        inherit_bit_visible = inherit_bit_test(fd, dev_info);

      if (!inherit_bit_visible)
      {
        entry_ptr = _get_sft_entry_ptr(fd);
        if (entry_ptr == 0)
        {
          /* The fd has already been validated, so reaching here means
             something is wrong with _get_sft_entry_ptr. */
          return -1;
        }


        dev_info = _farpeekw(_dos_ds, entry_ptr + 5);
      }


      if (flag & FD_CLOEXEC)
        dev_info |= _DEV_NO_INHERIT;
      else
        dev_info &= ~(_DEV_NO_INHERIT);


      if (inherit_bit_visible)
      {
        regs.x.ax = 0x4401;
        regs.x.bx = fd;
        regs.x.dx = dev_info;
        __dpmi_int(0x21, &regs);
      }
      else
        _farpokew(_dos_ds, entry_ptr + 5, dev_info);


      return 0;
    }


    case F_GETFL:
    {
      unsigned long sft_entry_ptr;
      unsigned char sft_open_flags;
      int flags = 0;

      /*
       * Use the data in the SFT.
       *
       * This seems to work on Windows 2000 and therefore probably works
       * on Windows XP. It does not work on Windows NT 4, where the file
       * will always appear to have been opened read-only. It's hard
       * to distinguish Windows 2000 and Windows NT 4 with an LFN TSR,
       * so we can't have a special case for Windows NT 4.
       */
      sft_entry_ptr = _get_sft_entry_ptr(fd);
      if (sft_entry_ptr)
      {
	/* Offset 2 in the SFT contains the open flags. */
	sft_open_flags = _farpeekw(_dos_ds, sft_entry_ptr + 2);

	switch(sft_open_flags & 0x7)
	{
	case 0: flags |= O_RDONLY; break;
	case 1: flags |= O_WRONLY; break;
	case 2: flags |= O_RDWR;   break;

	default:
	  break;
	}
      }
      else
      {
	/* If we've failed to get an SFT pointer, pretend that the file
	 * is read-write, since that's most likely to be correct. */
	flags |= O_RDWR;
      }

      /* Check the FILE_DESC_APPEND flag in __fd_properties. */
      if (__has_fd_properties(fd))
      {
	unsigned long fd_flags = __get_fd_flags(fd);

	if (fd_flags & FILE_DESC_APPEND)
	  flags |= O_APPEND;
      }

      return flags;
    }


    case F_SETFL:
    {
      unsigned char new_mode_bits;


      va_start (ap, cmd);
      new_mode_bits = va_arg(ap,int);
      va_end(ap);


      /* Allow removal of O_NONBLOCK, since DJGPP doesn't support it
         anyway.  */
      return (new_mode_bits == 0) ? 0 : -1;
    }


    case F_GETLK:
    case F_SETLK:
    case F_SETLKW:
    {
      struct flock *lock_req = NULL; /* shut up -Wall */
      struct flock64 lock_r64;
      int ret = -1, lcmd = -1;

      va_start (ap, cmd);
      lock_req = va_arg(ap, struct flock *);
      va_end (ap);

      switch (cmd)
      {
        case F_GETLK:   lcmd = F_GETLK64;  break;
        case F_SETLK:   lcmd = F_SETLK64;  break;
        case F_SETLKW:  lcmd = F_SETLKW64; break;
      }

      lock_r64.l_type   = lock_req->l_type;
      lock_r64.l_start  = lock_req->l_start;
      lock_r64.l_whence = lock_req->l_whence;
      lock_r64.l_len    = lock_req->l_len;

      ret = _fcntl_lk64(fd, lcmd, &lock_r64);

      lock_req->l_type   =         lock_r64.l_type;
      lock_req->l_start  = (off_t) lock_r64.l_start;
      lock_req->l_whence =         lock_r64.l_whence;
      lock_req->l_len    = (off_t) lock_r64.l_len;

      return ret;
    }


    case F_GETLK64:
    case F_SETLK64:
    case F_SETLKW64:
    {
      struct flock64 *lock_r64; /* shut up -Wall */
      int ret;

      va_start (ap, cmd);
      lock_r64 = va_arg(ap, struct flock64 *);
      va_end (ap);

      ret = _fcntl_lk64(fd, cmd, lock_r64);

      return ret;
    }
  }


  /* In case fcntl is called with an unrecognized command.  */
  errno = ENOSYS;
  return -1;
}
