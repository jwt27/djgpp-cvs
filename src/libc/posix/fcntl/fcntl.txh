@node fcntl, io
@findex fcntl
@subheading Syntax

@example
#include <fcntl.h>

int fcntl (int fd, int cmd, ...);
@end example

@subheading Description

This function performs the operation specified by @var{cmd} on
the file open on handle @var{fd}.  The following operations are defined
by the header @code{fcntl.h}:

@table @code
@item F_DUPFD
Returns a file handle that duplicates @var{fd} like @code{dup} does
(@pxref{dup}), except that @code{fcntl} also makes sure the returned
handle is the lowest available handle greater than or equal to the
integer value of the third argument.

@item F_GETFD
Get the @code{FD_CLOEXEC} close-on-exec (a.k.a.@: no-inherit) status of
@var{fd}.  If the returned value has its least-significant bit set, the
file will not be inherited by programs invoked by this process;
otherwise, the file will remain open in the child processes.

Note that only the first 20 handles can be passed to child processes by
DOS/Windows; handles beyond that cannot be inherited.  In addition, the
stub loader of the child DJGPP program will forcibly close handles 19
and 18 (since otherwise it will be unable to read the COFF executable
information and enter protected mode).  Therefore, the current
implementation always returns @code{FD_CLOEXEC} for handles 18 and
above.

For handles less than 18, the call will try to determine the status of
the @code{O_NOINHERIT} flag for that file and will return either
@code{FD_CLOEXEC} if the flag is set, or 0 if the flag is not set.  If
the status of the @code{O_NOINHERIT} flag cannot be determined, the call
will return -1, setting @code{errno} to @code{ENOSYS}.

The no-inherit bit can be set when the file is opened by using the
@code{O_NOINHERIT} in the open flags; see @ref{open}.

@item F_SETFD
Set or unset the close-on-exec flag for the handle @var{fd} using the
LSB of the integer value supplied as the third argument.  Since only the
first 20 handles are passed to child programs, and since the stub loader
of the child DJGPP program will forcibly close handles 19 and 18 (since
otherwise it will be unable to read the COFF executable information and
enter protected mode), the flag can only be set or unset on the first 18
handles.  Attempts to set the flag for handles 18 or above will always
return 0, and attempts to unset the flag for handles 18 or above will
always return -1, setting @code{errno} to @code{ENOSYS}.

For handles less than 18, the call will try to set or unset the
@code{O_NOINHERIT} flag for that file and will return 0 if the flag is
changed.  If the @code{O_NOINHERIT} flag cannot be changed, the call
will return -1, setting @code{errno} to @code{ENOSYS}.
  
@item F_GETFL
Get the open mode and status flags associated with the handle @var{fd}.
The flags are those supported by @code{open} and @code{creat} functions,
like @code{O_RDONLY}, @code{O_APPEND}, etc.

On Windows NT this cannot report the open mode correctly ---
@code{O_RDONLY} is always returned.

@item F_SETFL
Set the open mode and status flags associated with the handle @var{fd}.
This fails in all but one case, and sets @code{errno} to @code{ENOSYS},
since DOS and Windows don't allow changing the descriptor flags after
the file is open.

The one allowed case is for @code{O_NONBLOCK}, since DJGPP doesn't
support it anyway.  That is, calls using @code{F_SETFL} will fail for
all flag values @strong{except} @code{O_NONBLOCK}.

@example
#include <fcntl.h>

ret = fcntl(fd, F_SETFL, O_BINARY); /* This will fail, returning -1  */
                                    /* and setting errno to ENOSYS.  */

ret = fcntl(fd, F_SETFL, O_NONBLOCK); /* This will succeed          */
                                      /* returning 0.               */
@end example

@item F_GETLK
Return the lock structure that prevents obtaining the lock pointed to by
the third argument, or set the @code{l_type} field of the lock structure
to @code{F_UNLCK} if there is no obstruction.  Currently, only the
setting of the @code{l_type} field is provided.  This call will not
return values in the @code{struct flock} parameter identifying what lock
parameters prevent getting the requested lock, since there is no way to
obtain this information from DOS/Windows.  If the lock cannot be
obtained, -1 is returned and @code{errno} is set to the reason (which
will be one of @code{EINVAL}, @code{EBADF}, @code{EACCES} or
@code{ENOLCK}).

Locking of directories is not supported.

@item F_SETLK
Set or clear a file segment lock according to the structure pointed to
by the third argument.  The lock is set when @code{l_type} is
@code{F_RDLCK} (shared lock request) or @code{F_WRLCK} (exclusive lock
request), and the lock is cleared when @code{l_type} is @code{F_UNLCK}.
If the lock is already held, then this call returns -1 and sets
@code{errno} to @code{EACCES}.

The @code{F_RDLCK} value for requesting a read lock is always treated as
if it were @code{F_WRLCK} for a write lock.

This is because DOS/Win9x only supports one kind of lock, and it is the
exclusive kind.

Locking of directories is not supported.

@item F_SETLKW
Same as @code{F_SETLK}, but if the lock is blocked, the call will wait
(using @code{__dpmi_yield}, see @pxref{__dpmi_yield}) until it is
unblocked and the lock can be applied.  This call will @strong{never
exit} if the program making the call is the program which already owns
the lock.

Locking of directories is not supported.

@item F_GETLK64
@item F_SETLK64
@item F_SETLKW64
Each of these does exactly the same function as the non-"64" version,
but the third argument must be of type @code{struct flock64}, which
allows the @var{l_start} and @var{l_len} members to be @code{long long
int} values.  The current code will only use these @code{long long int}
values modulo @math{2^32}, which allows file locking positions up to 4
gigabytes minus 1.  True 64-bit file locking is @strong{not} supported.

The @code{struct flock64} members @var{l_start} and @var{l_len} are
declared to be of type @code{offset_t}, which is in turn typedef'ed to
be a @code{long long}.

Locking of directories is not supported.
@end table

This function can be hooked by the @dfn{Filesystem extensions}, see
@ref{File System Extensions}.  If you don't want this, and you are
calling @code{fcntl} with the @code{F_DUPFD} command, you should use
@code{dup2} instead, see @ref{dup2}.

@subheading Return Value

If an invalid or unsupported value is passed in @var{cmd}, or @var{fd}
is an invalid file handle, the function returns -1 and sets @code{errno}
to the appropriate value.  Unsupported values of @var{cmd} cause
@code{ENOSYS} to be stored in @code{errno}.  If @var{cmd} is
@code{F_DUPFD}, the function returns the new descriptor or -1 in case of
a failure.

Lock requests which specify the open file's current @code{EOF} position
as the value of @var{l_start} and zero as the @var{l_len} value will
fail, returning -1 with @code{errno} set to @code{EACCES}.


@subheading Portability

@port-note posix Contrary to Posix requirement, the handle returned by @code{F_DUPFD} shares the @code{FD_CLOEXEC} flag with @var{fd} (unless they are on different sides of the 20-handle mark), since DOS/Windows only maintain a single set of bits for all the handles associated with the same call to @code{open}.
@portability !ansi, posix


@subheading Example

@example
 /* Save the handle in a way that it won't be passed
    to child processes.  */
  int saved_fd = fcntl(fd, F_DUPFD, 20);

 /* Set an advisory lock for the whole file.  */
  struct flock flock;
  int retval, fd;

  flock.l_type = F_RDLCK;
  flock.l_whence = SEEK_SET;
  flock.l_start = flock.l_len = 0;
  errno = 0;
  retval = fcntl(fd, F_SETLK, &flock);

 /* Get the status of the lock we just obtained
    (should return -1 with errno == EACCES).  */
  errno = 0;
  retval = fcntl(fd, F_GETLK, &flock);

 /* Release the lock.  */
  errno = 0;
  flock.l_type = F_UNLCK;
  retval = fcntl(fd, F_SETLK, &flock);

 /* Get the status of the lock we just released
    (should return 0).  */
  errno = 0;
  flock.l_type = F_RDLCK;
  retval = fcntl(fd, F_GETLK, &flock);

 /* Try to set the O_BINARY flag on the open file
    (should return -1 with errno == ENOSYS).  */
  errno = 0;
  retval = fcntl(fd, F_SETFL, O_BINARY);

 /* Set the O_NONBLOCK flag on the open file
    (should return 0).  */
  errno = 0;
  retval = fcntl(fd, F_SETFL, O_NONBLOCK);

 /* Get the flags on the open file
    (always returns 0).  */
  errno = 0;
  retval = fcntl(fd, F_GETFL);

@end example
