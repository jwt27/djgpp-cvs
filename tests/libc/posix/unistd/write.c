#include <unistd.h>
#include <fcntl.h>

char buf[] = "line1\nline2\ncr\rcr\n";

int
main(void)
{
  int i = open("write.out", O_WRONLY|O_CREAT|O_TRUNC);
  write(i, buf, strlen(buf));
  close(i);
  return 0;
}
