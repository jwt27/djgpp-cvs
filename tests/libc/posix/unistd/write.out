line1
line2
cr
cr
